### PR TITLE
Refactor the thumbnail code

### PR DESCRIPTION
**Pull request by Snuffleupagus**
This PR rewrites part of the thumbnail code to achieve the following goals:
- Change the `ThumbnailView` implementation to be class-like.
- Refactor the `ThumbnailView` code to be more in-line with the corresponding `Page` code. (Even though the `Page` code has evolved quite a bit, the `Thumbnail` code hasn't usually followed suite.)
- Remove some dead code, and add a few memory related optimizations when discarding canvases.

To actually make it feasible to review this, I've divided the PR into multiple patches. The commit messages should _hopefully_ be clear enough to explain what is going on. Also, for some of the larger patches, using the `?w=1` URL flag should help.

I've made sure that at every point in the patch series, the code actually works. _However_, before landing this, I'm assuming that we might want to merge a few of the (smaller) commits together. But in order to make it somewhat easier to deal with any (hopefully non-existent) regressions, landing this as one big patch is probably not the best idea.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Review progress:

- [ ] Move PDFThumbnailViewer to its own file
- [ ] Fix function names in PDFThumbnailViewer
- [ ] Rename the thumbnail_view.js file
- [ ] Rename ThumbnailView to PDFThumbnailView and refactor it to be more class-like
- [ ] Refactor PDFThumbnailView to look more similar to PDFPageView
- [ ] Enable cancelling of thumbnail drawing
- [ ] Remove dead code from PDFThumbnailView

File: web/pdf_thumbnail_view.js [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+'use strict';
+
+var THUMBNAIL_WIDTH = 98; // px
+var THUMBNAIL_CANVAS_BORDER_WIDTH = 1; // px
+
+/**
+ * @typedef {Object} PDFThumbnailViewOptions
+ * @property {HTMLDivElement} container - The viewer element.
+ * @property {number} id - The thumbnail unique ID (normally its number).
```
Comment (timvandermeij) [inline web/pdf_thumbnail_view.js]: Nit: The thumbnail*'s* unique ID

Comment (Snuffleupagus) [inline web/pdf_thumbnail_view.js]: Fixed.

**Comment by yurydelendik** [comment]
file and symbols dependencies look good

**Comment by timvandermeij** [comment]
Awesome work!

**Pull request merged**